### PR TITLE
php.packages: Make packages overridable

### PR DESCRIPTION
**Pull request by jtojnar**
###### Motivation for this change
This will allow patching extensions without having to manually having to patch all occurrences in the dependency tree (like I did in https://github.com/fossar/selfoss/commit/cb22131a29ae25ff94f044d5079044dfb22b3a2b), see [clean-up PR](https://github.com/fossar/selfoss/pull/1235) depending on this PR.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by aanderse** [comment]
> Unfortunately, Nixpkgs repository does not contain old PHP versions
but it is trivial with Nix to build them ourselves. We will be able to
use Cachix to avoid rebuilding them over and over again.

It sounds like you're going to publish this for broad use? I have a few internal applications I need to support still running `5.6` and this would be really nice... Would you mind sending me a ping when this is available?

**Comment by jtojnar** [comment]
@aanderse It is already in the selfoss repository (see the links in the OP). Though I hope to remove PHP 5.6 as soon as I manage to release selfoss 2.19, which is the last version I promised PHP 5 support for.

**Comment by aanderse** [comment]
@jtojnar there isn't a binary cache yet, though, is there? Thanks! Awesome work.

**Comment by jtojnar** [comment]
@aanderse Yes, there is. Turns out Cachix is amazing: `cachix use fossar`

**Comment by FRidh** [comment]
Related PR for emacs https://github.com/NixOS/nixpkgs/pull/107152.

**Comment by FRidh** [comment]
Wrote a draft for an RFC on this topic https://github.com/NixOS/rfcs/pull/83.

File: pkgs/development/interpreters/php/generic.nix [hunk 1]
```diff
@@ -13,6 +13,7 @@ let
     , version
     , sha256
     , extraPatches ? []
+    , packageOverrides ? (final: prev: {})
```
[the remark this answers is not in the document]
Comment (jtojnar) [reply pkgs/development/interpreters/php/generic.nix:16]: It does actually compose:

```nix
let
  pkgs = import ./. { };

  idOverlay = final: prev: {};
  composeOverlays = old: new:
    final: prev: new final (old final prev);

  oldPhp = pkgs.php.override {
    packageOverrides = final: prev: {
      extensions = prev.extensions // {
        mysqlnd = prev.extensions.mysqlnd.overrideAttrs (attrs: {
          foo = "foo";
        });
      };
    };
  };
in
  oldPhp.override (oldOverrides: {
    packageOverrides = composeOverlays (oldOverrides.packageOverrides or idOverlay) (final: prev: {
      extensions = prev.extensions // {
        mysqlnd = prev.extensions.mysqlnd.overrideAttrs (attrs: {
          foo = prev.extensions.mysqlnd.foo or "" + "bar";
        });
      };
    });
  })
```

But yeah, it is a mouthful.

**Comment by jtojnar** [comment]
Merging as is for now, will update it to match the RFC once we figure it out.

**Review by etu — COMMENTED** [review]
I approve of these changes since it doesn't make sense to wait around for the RFC process to happen. This brings us benefits right now, and we can match the RFC API later on.

**Comment by jtojnar** [comment]
@aanderse I have extracted the PHP expressions into a separate repository: https://github.com/fossar/nix-phps

**Comment by aanderse** [comment]
@jtojnar thank you very much! That is great! ✨

**Pull request merged**